drivers/dma/dw-cdma/dw-cdma-platform.c

```diff
@@ -0,0 +1,664 @@
+#include <linux/module.h>
+#include <linux/of_dma.h>
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+#include <linux/delay.h>
+#include <asm/cluster.h>
+#include "dw-cdma.h"
+#include "../dmaengine.h"
+#include "../virt-dma.h"
+
+
+static struct cdma_chip chip;
+
+static void cdma_enable_ch(u32 ch)
+{
+	struct cdma_s_stat_t stat;
+
+	stat.r = 0; /* Reset channel */
+	stat.e = 1; /* Enable channel */
+	WRITE_AUX(DMA_S_STATC_AUX(ch), stat);
+
+	do {
+		READ_BCR(DMA_S_STATC_AUX(ch), stat);
+		if (stat.st == CDMA_IDLE || stat.st == CDMA_BUSY)
+			break;
+	} while (1);
+}
+
+static void cdma_disable_ch(u32 ch)
+{
+	struct cdma_s_stat_t stat;
+
+	stat.r = 0; /* Reset channel */
+	stat.e = 0; /* Disable channel */
+	WRITE_AUX(DMA_S_STATC_AUX(ch), stat);
+
+	do {
+		READ_BCR(DMA_S_STATC_AUX(ch), stat);
+	} while (stat.st == CDMA_BUSY);
+}
+
+static void cdma_reset_ch(u32 ch)
+{
+	struct cdma_s_stat_t stat;
+
+	stat.r = 1; /* Reset channel */
+	stat.e = 0; /* Enable channel */
+	WRITE_AUX(DMA_S_STATC_AUX(ch), stat);
+
+	do {
+		READ_BCR(DMA_S_STATC_AUX(ch), stat);
+	} while (stat.st == CDMA_BUSY);
+}
+
+static inline int cdma_get_bit_done(u32 handle)
+{
+	u32 a, o, x;
+
+	a = handle >> 5;
+	o = 1 << (handle & 0x1f);
+	READ_BCR(DMA_S_DONESTATD_AUX(a), x);
+	x = x & o;
+	return x != 0;
+}
+
+static inline void cdma_clear_bit_done(u32 handle)
+{
+	u32 a, x;
+
+	a = handle >> 5;
+	x = 1 << (handle & 0x1f);
+	WRITE_AUX(DMA_S_DONESTATD_CLR_AUX(a), x);
+}
+
+/* Get descriptor handle of the last started DMA */
+static inline u32 cdma_get_handle(void)
+{
+	struct cdma_c_stat_t stat;
+	u32 handle;
+
+	do {
+		READ_BCR(DMA_C_STAT_AUX, stat);
+	} while (stat.b);
+
+	READ_BCR(DMA_C_HANDLE_AUX, handle);
+	return handle;
+}
+
+static int free_hw_desc_list(struct list_head *list_head)
+{
+	struct cdma_hw_desc *hw_desc, *hw_desc_temp;
+
+	if (list_empty(list_head))
+		return 1;
+
+	list_for_each_entry_safe(hw_desc, hw_desc_temp, list_head, list) {
+		list_del(&hw_desc->list);
+		kfree(hw_desc);
+	}
+	return 0;
+}
+
+static void vchan_desc_put(struct virt_dma_desc *vdesc)
+{
+	struct c_dma_desc *desc = vd_to_cdma_desc(vdesc);
+	unsigned long flags;
+	int done_flg;
+
+	spin_lock_irqsave(&desc->chan->vc.lock, flags);
+	done_flg = free_hw_desc_list(&desc->hw_desc_alloc_list);
+	done_flg &= free_hw_desc_list(&desc->hw_desc_issued_list);
+	done_flg &= !free_hw_desc_list(&desc->hw_desc_finished_list);
+	spin_unlock_irqrestore(&desc->chan->vc.lock, flags);
+
+	if (!done_flg)
+		dev_warn(chan2dev(desc->chan), "WARN. Put an active descriptor.\n");
+
+	kfree(desc);
+}
+
+static void cdma_setup(struct cdma_chip *chip)
+{
+	struct cdma_c_ctrl_t ctrl_c;
+	struct cdma_s_ctrl_t ctrl_s;
+	struct cdma_s_prioc_t prio_ch;
+	struct dw_cluster_dma *dw = chip->dw;
+	struct c_dma_chan *chan;
+	int desc_per_ch;
+	int ii;
+
+	/* Define number of descriptors per each chanel */
+	desc_per_ch = dw->nr_descr / dw->nr_channels;
+
+	for (ii = 0; ii < dw->nr_channels; ii++) {
```

Comment:
We can just define counter in the loop:
```suggestion
	for (int ii = 0; ii < dw->nr_channels; ii++) {
```